File: src/julia/src/Monoidal.jl
# Monoidal category -- a category equipped with a tensor product and unit.
#
# Mathematical definition (nLab: monoidal+category):
# A monoidal category (C, tensor, I, alpha, lambda, rho) consists of:
# - A category C
# - A bifunctor tensor : C x C -> C (the tensor product)
# - An object I in C (the monoidal unit)
# - A natural isomorphism alpha_{A,B,C} : (A tensor B) tensor C ~-> A tensor (B tensor C)
# - A natural isomorphism lambda_A : I tensor A ~-> A (left unitor)
# - A natural isomorphism rho_A : A tensor I ~-> A (right unitor)
#
# subject to the pentagon and triangle coherence axioms.
#
# Pentagon axiom:
#   (id_A bimap alpha_{B,C,D}) . alpha_{A, B tensor C, D} . (alpha_{A,B,C} bimap id_D)
#   = alpha_{A,B, C tensor D} . alpha_{A tensor B, C, D}
#
# Triangle axiom:
#   (id_A bimap lambda_B) . alpha_{A,I,B} = rho_A bimap id_B
#
# Design decision: @theory extending ThCategory with full coherence data.
# Pentagon and triangle axioms are stated in comments (their nested term
# depth may exceed GATlab's equation parser) and verified in tests.

using GATlab

"""
    ThMonoidalCategory

The theory of monoidal categories, extending ThCategory.

Additional operations:
- `otimes(a, b)` -- tensor product on objects
- `otimes(f, g)` -- tensor product on morphisms (bifunctoriality)
- `munit()` -- the monoidal unit object
- `associator(a, b, c)` -- alpha_{a,b,c} : (a otimes b) otimes c -> a otimes (b otimes c)
- `associator_inv(a, b, c)` -- alpha^{-1}
- `left_unitor(a)` -- lambda_a : munit otimes a -> a
- `left_unitor_inv(a)` -- lambda^{-1}_a
- `right_unitor(a)` -- rho_a : a otimes munit -> a
- `right_unitor_inv(a)` -- rho^{-1}_a

Axioms include associator/unitor roundtrips, bifunctoriality, pentagon, and triangle.

nLab: monoidal+category.
"""
@theory ThMonoidalCategory <: ThCategory begin
    # Tensor product on objects
    otimes(a::Ob, b::Ob)::Ob

    # Tensor product on morphisms (bifunctoriality)
    otimes(f::Hom(a, b), g::Hom(c, d))::Hom(otimes(a, c), otimes(b, d)) ⊣
        [a::Ob, b::Ob, c::Ob, d::Ob]

    # Monoidal unit
    munit()::Ob

    # Associator and its inverse
    associator(a::Ob, b::Ob, c::Ob)::Hom(otimes(otimes(a, b), c), otimes(a, otimes(b, c)))
    associator_inv(a::Ob, b::Ob, c::Ob)::Hom(otimes(a, otimes(b, c)), otimes(otimes(a, b), c))

    # Left unitor and its inverse
    left_unitor(a::Ob)::Hom(otimes(munit(), a), a)
    left_unitor_inv(a::Ob)::Hom(a, otimes(munit(), a))

    # Right unitor and its inverse
    right_unitor(a::Ob)::Hom(otimes(a, munit()), a)
    right_unitor_inv(a::Ob)::Hom(a, otimes(a, munit()))

    # --- Roundtrip axioms (isomorphism conditions) ---

    # Associator roundtrip
    compose(associator(a, b, c), associator_inv(a, b, c)) == id(otimes(otimes(a, b), c)) ⊣
        [a::Ob, b::Ob, c::Ob]
    compose(associator_inv(a, b, c), associator(a, b, c)) == id(otimes(a, otimes(b, c))) ⊣
        [a::Ob, b::Ob, c::Ob]

    # Left unitor roundtrip
    compose(left_unitor(a), left_unitor_inv(a)) == id(otimes(munit(), a)) ⊣ [a::Ob]
    compose(left_unitor_inv(a), left_unitor(a)) == id(a) ⊣ [a::Ob]

    # Right unitor roundtrip
    compose(right_unitor(a), right_unitor_inv(a)) == id(otimes(a, munit())) ⊣ [a::Ob]
    compose(right_unitor_inv(a), right_unitor(a)) == id(a) ⊣ [a::Ob]

    # --- Tensor bifunctoriality axioms ---

    # Tensor preserves identity
    otimes(id(a), id(b)) == id(otimes(a, b)) ⊣ [a::Ob, b::Ob]

    # Tensor preserves composition
    compose(otimes(f1, g1), otimes(f2, g2)) == otimes(compose(f1, f2), compose(g1, g2)) ⊣
        [a::Ob, b::Ob, c::Ob, d::Ob, e::Ob, f_::Ob,
         f1::Hom(a, b), f2::Hom(b, c), g1::Hom(d, e), g2::Hom(e, f_)]

    # --- Pentagon axiom ---
    # compose(otimes(id(a), associator(b,c,d)),
    #         compose(associator(a, otimes(b,c), d),
    #                 otimes(associator(a,b,c), id(d))))
    # == compose(associator(a, b, otimes(c,d)), associator(otimes(a,b), c, d))
    #
    # Verified in tests; nested term depth may exceed GATlab's equation parser.

    # --- Triangle axiom ---
    # compose(otimes(id(a), left_unitor(b)), associator(a, munit(), b))
    # == otimes(right_unitor(a), id(b))
    #
    # Verified in tests.
end

File: src/haskell/src/Cat/Monoidal.hs
-- | Monoidal category -- a category equipped with a tensor product and unit.
--
-- Mathematical definition (nLab: monoidal+category):
-- A monoidal category \((\mathcal{C}, \otimes, I, \alpha, \lambda, \rho)\)
-- consists of:
--
-- * A category \(\mathcal{C}\)
-- * A bifunctor \(\otimes : \mathcal{C} \times \mathcal{C} \to \mathcal{C}\) (the tensor product)
-- * An object \(I \in \mathcal{C}\) (the monoidal unit)
-- * A natural isomorphism \(\alpha_{A,B,C} : (A \otimes B) \otimes C \xrightarrow{\sim} A \otimes (B \otimes C)\) (the associator)
-- * A natural isomorphism \(\lambda_A : I \otimes A \xrightarrow{\sim} A\) (the left unitor)
-- * A natural isomorphism \(\rho_A : A \otimes I \xrightarrow{\sim} A\) (the right unitor)
--
-- subject to the pentagon and triangle coherence axioms.
--
-- __Pentagon axiom__: For all objects \(A, B, C, D\), the following diagram commutes:
--
-- @
--                          alpha_{A,B,C} * id_D
--   ((A*B)*C)*D  -------------------------------->  (A*(B*C))*D
--        |                                               |
--        | alpha_{A*B,C,D}                               | alpha_{A,B*C,D}
--        v                                               v
--   (A*B)*(C*D)                                    A*((B*C)*D)
--        |                                               |
--        | alpha_{A,B,C*D}                               | id_A * alpha_{B,C,D}
--        v                                               v
--   A*(B*(C*D))  <================================  A*(B*(C*D))
-- @
--
-- That is:
-- @(id_A `bimap` alpha_{B,C,D}) . alpha_{A,B*C,D} . (alpha_{A,B,C} `bimap` id_D)@
-- @= alpha_{A,B,C*D} . alpha_{A*B,C,D}@
--
-- __Triangle axiom__: For all objects \(A, B\), the following diagram commutes:
--
-- @
--   (A*I)*B ---alpha_{A,I,B}---> A*(I*B)
--       \                         /
--        \                       /
--    rho_A * id_B         id_A * lambda_B
--          \                   /
--           v                 v
--               A * B
-- @
--
-- That is:
-- @(id_A `bimap` lambda_B) . alpha_{A,I,B} = rho_A `bimap` id_B@
--
-- This module provides the data track encoding only (records, not typeclasses).
-- The associator and unitors are stored as rank-2 component families rather than
-- as 'NatIso' values, since the associator is a natural isomorphism between
-- /trifunctors/ \(\mathcal{C}^3 \to \mathcal{C}\), which does not fit the
-- shape of 'NatIso' (parameterized over unary type constructors).
--
-- Laws (documented here, tested via QuickCheck in "Cat.MonoidalSpec"):
--
-- [Associator roundtrip]
--   @compose monAssocBwd monAssocFwd = id@
--   @compose monAssocFwd monAssocBwd = id@
--
-- [Left unitor roundtrip]
--   @compose monLeftUnitorBwd monLeftUnitorFwd = id@
--   @compose monLeftUnitorFwd monLeftUnitorBwd = id@
--
-- [Right unitor roundtrip]
--   @compose monRightUnitorBwd monRightUnitorFwd = id@
--   @compose monRightUnitorFwd monRightUnitorBwd = id@
--
-- [Associator naturality]
--   For all @f@, @g@, @h@:
--   @monAssocFwd . bimap (bimap f g) h = bimap f (bimap g h) . monAssocFwd@
--
-- [Pentagon]
--   @compose (bimap id monAssocFwd) (compose monAssocFwd (bimap monAssocFwd id))@
--   @= compose monAssocFwd monAssocFwd@
--
-- [Triangle]
--   @compose (bimap id monLeftUnitorFwd) monAssocFwd = bimap monRightUnitorFwd id@
--
-- nLab: monoidal+category.
module Cat.Monoidal
  ( MonoidalData(..)
  ) where

import Prelude hiding (id, (.))
import Data.Kind (Type)

import Cat.Category (CategoryData(..))
import Cat.Bifunctor (BifunctorData(..))

-- | A monoidal category reified as a record.
--
-- The type parameters are:
--
-- * @hom@ -- the morphism type of the underlying category (kind @Type -> Type -> Type@)
-- * @tensor@ -- the tensor product on objects (kind @Type -> Type -> Type@)
-- * @unit@ -- the monoidal unit (kind @Type@)
--
-- The coherence isomorphisms (associator, left unitor, right unitor) are stored
-- as forward and backward component families. This is the correct representation
-- for the data track: the associator is a natural isomorphism between trifunctors,
-- and the unitors are natural isomorphisms between endofunctors, but we store their
-- components directly rather than wrapping them in 'NatIso'.
--
-- nLab: monoidal+category.
data MonoidalData
  (hom :: Type -> Type -> Type)
  (tensor :: Type -> Type -> Type)
  (unit :: Type)
  = MonoidalData
  { monCat :: CategoryData hom
    -- ^ The underlying category.
  , monTensor :: BifunctorData hom hom hom tensor
    -- ^ The tensor product bifunctor \(\otimes : \mathcal{C} \times \mathcal{C} \to \mathcal{C}\).
  , monAssocFwd :: forall a b c. hom (tensor (tensor a b) c) (tensor a (tensor b c))
    -- ^ The forward component of the associator:
    -- \(\alpha_{A,B,C} : (A \otimes B) \otimes C \to A \otimes (B \otimes C)\).
  , monAssocBwd :: forall a b c. hom (tensor a (tensor b c)) (tensor (tensor a b) c)
    -- ^ The backward component of the associator:
    -- \(\alpha^{-1}_{A,B,C} : A \otimes (B \otimes C) \to (A \otimes B) \otimes C\).
  , monLeftUnitorFwd :: forall a. hom (tensor unit a) a
    -- ^ The forward component of the left unitor:
    -- \(\lambda_A : I \otimes A \to A\).
  , monLeftUnitorBwd :: forall a. hom a (tensor unit a)
    -- ^ The backward component of the left unitor:
    -- \(\lambda^{-1}_A : A \to I \otimes A\).
  , monRightUnitorFwd :: forall a. hom (tensor a unit) a
    -- ^ The forward component of the right unitor:
    -- \(\rho_A : A \otimes I \to A\).
  , monRightUnitorBwd :: forall a. hom a (tensor a unit)
    -- ^ The backward component of the right unitor:
    -- \(\rho^{-1}_A : A \to A \otimes I\).
  }
